### PR TITLE
Pagination component

### PR DESCRIPTION
**Pull request by Cyanideankst**

### CODE REVIEW

**Comment by VilppeRiskidev** [comment]
CSS stuff is (intentionally) a bit gross. It can be refactored once we figure out "the new way".
Didn't know yet what to do with "New Purple"-color, it is currently hard-coded. What's the situation with that new color, do we start using it / add it to global.css.....

File: packages/cyberstorm/src/components/componentStyles/Pagination.module.css [hunk 1]
```diff
@@ -0,0 +1,20 @@
+.root {
+  display: flex;
+  flex-direction: row;
+  gap: 4px;
+  align-items: center;
+  justify-content: center;
+
+  width: 407px;
+  height: 36px;
+  padding: 0;
```
Comment (gkurck) [inline packages/cyberstorm/src/components/componentStyles/Pagination.module.css:10]: I'm guessing this is here to eliminate the browser default padding for `<ul>`? This is also included in the `reset.css`, which carpet bombs everything to have `margin:0; padding:0;`. Of course we can't rely on us alwasys having that reset, and as such we should collectively have a chat about the feasibility of using `appearance:none` and `all:unset` on our components where applicable.

File: packages/cyberstorm/src/components/componentStyles/Pagination.module.css [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+.disabled {
+  pointer-events: none;
+}
+
+.disabled:hover {
+  background-color: transparent;
+  cursor: default;
+}
```
Comment (gkurck) [inline packages/cyberstorm/src/components/componentStyles/Pagination.module.css:20]: These are redundant, as the rule `pointer-events:none` in the previous selector already render the elements unhoverable.

**Review by gkurck — CHANGES_REQUESTED** [review]
Good stuff, added a few comments related to functionality, styling and semantics/accessibility.

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { ComponentStory, ComponentMeta } from "@storybook/react";
+import { Pagination } from "@thunderstore/cyberstorm";
+import React, { useState } from "react";
+const meta = {
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:4]: Just to be sure that you want to have empty line above this right?

File: packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+.button__default:hover {
+  --bg-color: var(--color-surface--3);
+}
+
+.button__default_selected {
+  --bg-color: #40328b;
+}
+
+.button__default_selected:hover {
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css:37]: Roger, Deleted that.

File: packages/cyberstorm/src/components/componentStyles/Pagination.module.css [hunk 5]
```diff
@@ -0,0 +1,20 @@
+.root {
+  display: flex;
+  flex-direction: row;
+  gap: 4px;
+  align-items: center;
+  justify-content: center;
+
+  width: 407px;
+  height: 36px;
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/componentStyles/Pagination.module.css:9]: Done.

File: packages/cyberstorm/src/components/componentStyles/Pagination.module.css (hunk 2 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/componentStyles/Pagination.module.css:20]: Aight, deleted these so we don't have unnecessary things hanging 👍

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    component size which we do not want
+  */
+  const shouldShowLeftDots = leftSiblingIndex > 2;
+  const shouldShowRightDots = rightSiblingIndex < totalPageCount - 2;
+
+  const firstPageIndex = 1;
+  const lastPageIndex = totalPageCount;
+
+  if (!shouldShowLeftDots && shouldShowRightDots) {
+    const leftItemCount = 3 + 2 * siblingCount;
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:51]: Why not just `const leftRange = range(1, 5);`?

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    const leftRange = range(1, leftItemCount);
+
+    return [...leftRange, DOTS, totalPageCount];
+  }
+
+  if (shouldShowLeftDots && !shouldShowRightDots) {
+    const rightItemCount = 3 + 2 * siblingCount;
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:58]: Why not just `const rightRange = range(totalPageCount - 4, totalPageCount);`?

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  if (shouldShowLeftDots && shouldShowRightDots) {
+    const middleRange = range(leftSiblingIndex, rightSiblingIndex);
+    return [firstPageIndex, DOTS, ...middleRange, DOTS, lastPageIndex];
+  }
+  return [];
+};
+
+export interface PaginationProps {
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:73]: Placing `PaginationProps` and `Pagination` between two helper functions seems odd. Maybe they should be the first things in this file after imports?

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+const mapPageNumbers = (
+  paginationRange: Array<number | string>,
+  onPageChange: Dispatch<SetStateAction<number>>,
+  currentPage: number
+) => {
+  return paginationRange.map((pageNumber, index: number) => {
+    if (pageNumber === DOTS) {
+      return <PaginationButton key={index} label="..." />;
+    }
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:115]: * The buttons should have some special className or prop which can be used to disable `cursor: pointer` (and maybe the hover effect too), since clicking them does nothing.
  * Edit: alternatively this could be done in `PaginationButton` by checking if `onClick` is undefined?
* Key could be ```key={`dots-${index}`}``` (React doesn't like plain indexes as keys.)
* Label should be `label={DOTS}`

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+    return (
+      <PaginationButton
+        key={index}
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:119]: ```key={`page-${index}´}```?

File: packages/cyberstorm/src/index.ts [hunk 11]
```diff
@@ -1,7 +1,9 @@
 export * from "./components/Button";
+export * from "./components/PaginationButton";
```
Comment (anttimaki) [inline packages/cyberstorm/src/index.ts:2]: Seems someone has already messed up the alphabetical ordering here, but could you please put these two new components right next to each other, e.g. after `PackageFlag`? :ocd:

**Review by anttimaki — CHANGES_REQUESTED** [review]

**Comment by anttimaki** [comment]
Also, by default GitHub creates the title of PR based on the branch name. You can edit it when creating the PR, as well as afterwards. "Ts 1375" isn't very descriptive when one browses the PR list. "TS-1375: Pagination" or "Add Cyberstorm Pagination component" would be better.

File: packages/cyberstorm/src/index.ts (hunk 11 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/index.ts:2]: :ocd: feels awful. Arranged them to be in alphabetical order.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 9 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:115]: Had to dig deeper to understand why React didn't like it but now I don't like them either. 
Will fix it to match your example 👍

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+  disabled?: boolean;
+}
+
+export const Pagination: React.FC<PaginationProps> = (props) => {
+  const { currentPage, onPageChange, totalCount, pageSize, disabled } = props;
+
+  const paginationRange = usePagination(totalCount, pageSize, currentPage);
+
+  return (
+    <ul className={`${styles.root} ${disabled ? styles.disabled : ""}`}>
```
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/Pagination.tsx:87]: To correct myself, `<nav>` already implies it's purpose in the name, so there's no need to communicate it twice. `role="navigation"` could be attached to the `<ul>` if we weren't using `<nav>` to wrap everything in, but using the native, semantic elements is always better than adding semantics to other elements with ARIA.

File: packages/cyberstorm/src/components/PaginationButton.tsx [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  return (
+    <button
+      type="button"
+      className={`${styles.root} ${getStyle(colorScheme, isSelected)}`}
+      onClick={onClick}
+    >
+      {leftIcon}
+      {label ? <div className={styles.label}>{label}</div> : null}
+      {rightIcon}
+    </button>
```
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/PaginationButton.tsx:25]: To correct myself, while `aria-current="true"` is perfectly valid, more accurate value would be `aria-current="page"`, since that's what the item indicates - the current page in the pagination list we're on.

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+const Template: ComponentStory<typeof Pagination> = (args) => {
+  const [value, setValue] = useState(args.defaultPage ?? null);
+  args.onPageChange = setValue;
+  args.currentPage = value;
+  delete args.defaultPage;
+
+  return (
+    <div style={{ display: "flex" }}>
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:23]: Yup, didn't give second thoughts style. 
Was there some kinda consensus on other areas how or should we show these? @VilppeRiskidev

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+      <div style={{ color: "white" }}>Value in state: {value}</div>
+      <Pagination {...args} />
+    </div>
+  );
+};
+
+const ReferencePagination = Template.bind({});
+ReferencePagination.args = {
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:31]: I don't think so. Changed it to match your suggestion 👌

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+import { PaginationButton } from "./PaginationButton";
+
+const DOTS = "...";
+
+const range = (start: number, end: number) => {
+  const length = end - start + 1;
+  return Array.from({ length }, (_, idx) => idx + start);
+};
+
+const usePagination = (
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:14]: Changed it to match getButtonLabels 👍

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+  totalCount: number,
+  pageSize: number,
+  currentPage: number
+) => {
+  const siblingCount = 1;
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:19]: It refers how many page numbers we show before/after current page button.

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+  // Pages count is determined as siblingCount + firstPage + lastPage + currentPage + 2*DOTS
+  const totalPageNumbers = siblingCount + 5;
+
+  /*
+    If the number of pages is less than the page numbers we want to show in our
+    paginationComponent, we return the range [1..totalPageCount]
+  */
+  if (totalPageNumbers >= totalPageCount) {
+    return range(1, totalPageCount);
+  }
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:31]: siblingCount was intended to be able changed if we want to add more of those buttons on some cases. Wasn't sure tho was this wanted functionality  on second thought.

maxButtons do sound much better and will change it to that!

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+
+  const leftSiblingIndex = Math.max(currentPage - siblingCount, 1);
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:33]: I do agree that yours might be bit more intuitive way to do it so will change it to match that.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 10 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:119]: Did you mean like this key={`page-${index}`} ? 

That last back tick must have nudged over when sipping some coffee ☕ 
Did change it like that since keys on react..

File: packages/cyberstorm/src/components/componentStyles/Pagination.module.css (hunk 1 above)
Comment (gkurck) [reply packages/cyberstorm/src/components/componentStyles/Pagination.module.css:8]: You can leave `padding:0;` here for now as it's the desired result to reset such browser defaults for `<ul>` anyway. The ideas I mentioned shouldn't be made a blocker right now.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:87]: Ok, tried to understand what is correct way to do it on this so lets see if I got it correctly 👌

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (VilppeRiskidev) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:23]: No consensus. Do what looks best case by case.

File: packages/cyberstorm/src/components/PaginationButton.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/PaginationButton.tsx:25]: Yup, I think I got it now 👌

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+  ...defaultArgs,
+  defaultPage: 3,
+};
+
+const DisabledPagination = Template.bind({});
+DisabledPagination.args = {
+  ...defaultArgs,
+  disabled: true,
+};
+
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:46]: Fixed it not to render next/prev buttons when totalCount is 0.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 6 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:51]: siblingCount is now passable parameter on the pagination component.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 7 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:58]: siblingCount is now passable parameter on the pagination component.

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+      />
+
+      {mapPageNumbers(paginationRange, onPageChange, currentPage)}
+
+      <PaginationButton
+        onClick={() => onPageChange(currentPage + 1)}
+        label="Next"
+        rightIcon={<FontAwesomeIcon fixedWidth icon={faArrowRight} />}
+      />
+    </ul>
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:101]: This should be fixed now. 👍

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+) => {
+  // maxButtons is determined as siblingCount + firstPage + lastPage + currentPage + 2*DOTS
+  const maxButtons = siblingCount + 5;
+
+  // Show a button for each page if there's just a few of them.
+  if (maxButtons >= totalPageCount) {
+    return range(1, totalPageCount);
+  }
+
+  const prevPageNumber = Math.max(currentPage - siblingCount, 1);
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:32]: With all the other refactorings done here, I'm not sure these names are very good anymore. E.g. with `siblingCount` 3 and `currentPage` 3, `prevPageNumber` will be 1 and `nextPageNumber` 6, while one would probably expect 2 and 4. Would `leftmostSibling` and `rightmostSibling` make sense?

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+export const Pagination: React.FC<PaginationProps> = (props) => {
+  const {
+    currentPage,
+    onPageChange,
+    totalCount,
+    pageSize,
+    disabled,
+    siblingCount = 1,
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:67]: `disabled` and `siblingCount` are both optional props, but the former has the default value defined in `defaultProps` and the latter here. For consistency, define both defaults in one places.

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+  } = props;
+  const totalPageCount = Math.ceil(totalCount / (pageSize > 0 ? pageSize : 1));
+  const paginationRange = getButtonLabels(
+    totalPageCount,
+    currentPage,
+    siblingCount
+  );
+
+  function decreaseCurrentPage(currentPage: number) {
+    const y = currentPage - 1;
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:77]: Nitpick: while the variable name has little meaning when the scope where it's used is whole two lines, I think renaming it to `newPage` oslt might improve readability? This is not a hard-and-fast rule btw, the issue should be judged case by case.

File: packages/cyberstorm/src/components/Pagination.tsx [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+    return y > 0 ? y : 1;
+  }
+
+  function increaseCurrentPage(currentPage: number) {
+    const y = currentPage + 1;
+    return y <= totalPageCount ? y : totalPageCount;
+  }
+
+  if (totalCount === 0) {
+    return <></>;
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/Pagination.tsx:87]: Common practice is to return `null` if you don't want to render anything.

File: packages/cyberstorm/src/components/PaginationButton.tsx [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+    ariaLabel,
+    label,
+    leftIcon,
+    rightIcon,
+    colorScheme,
+    onClick,
+    isSelected,
+  } = props;
+  return (
+    <button
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/PaginationButton.tsx:27]: IIRC it was mentioned in the weekly that this should maybe be a `<a>` element instead. Although I guess it depends on what clicking the component actually does? I guess it's quite safe bet to say it only navigates (i.e. it should be `<a>`), but whoever uses the component might make it also change the state or whatever (i.e. it should be <button>). We have no control over that.

TL;DR: I've no strong opinions about this issue.

File: packages/cyberstorm/src/components/PaginationButton.tsx [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+      onClick={onClick}
+    >
+      {leftIcon}
+      {label ? <div className={styles.label}>{label}</div> : null}
+      {rightIcon}
+    </button>
+  );
+};
+
+PaginationButton.defaultProps = { colorScheme: "default" };
```
Comment (anttimaki) [inline packages/cyberstorm/src/components/PaginationButton.tsx:41]: All props are optional but only one has default value.

File: packages/cyberstorm/src/index.ts [hunk 28]
```diff
@@ -3,5 +3,8 @@ export * from "./components/LinkingProvider";
 export * from "./components/Links";
 export * from "./components/MetaItem";
 export * from "./components/PackageFlag";
-export * from "./components/Tag";
 export * from "./components/PackageCard";
+export * from "./components/Pagination";
+export * from "./components/PaginationButton";
+export * from "./components/Tag";
+export * from "./utils";
```
Comment (anttimaki) [inline packages/cyberstorm/src/index.ts:10]: IMO the utils should be only for the internal use within this package, and shouldn't be exported elsewhere.

File: packages/cyberstorm/src/utils/index.ts [hunk 29]
```diff
@@ -0,0 +1 @@
+export * from "./array";
```
Comment (anttimaki) [inline packages/cyberstorm/src/utils/index.ts:1]: If the utils are only for internal use, this file makes little sense.

**Review by anttimaki — CHANGES_REQUESTED** [review]
Note that some of the comments are not blockers, so if you feel they require lots of work and have little impact, feel free to argue against doing them.

Also note that the PR currently has merge conflicts which should be resolved.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/Pagination.tsx]: > Should the prev/next buttons be hidden if already on the first/last page? Or just disabled?

I'd say they should be hidden.

File: packages/cyberstorm/src/components/PaginationButton.tsx (hunk 26 above)
Comment (gkurck) [reply packages/cyberstorm/src/components/PaginationButton.tsx:27]: For reference, here's [my original comment about the this](https://github.com/thunderstore-io/thunderstore-ui/pull/604#discussion_r1081837678)

File: packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css [hunk 30]
```diff
@@ -0,0 +1,39 @@
+.root {
+  display: flex;
+  flex: none;
+  flex-direction: row;
+  gap: var(--gap-3xs);
+  align-items: center;
+  justify-content: center;
+  min-width: 36px;
+  min-height: 36px;
```
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css]: The notion about the min vs hard values still applies. Since both the prev/next buttons and the page numbers use the same component, you can do this in a couple of different ways; applying the appropriate css rules/class to the root or the label.

File: packages/cyberstorm/src/components/PaginationButton.tsx [hunk 31]
```diff
@@ -0,0 +1,56 @@
...
+    leftIcon,
+    rightIcon,
+    colorScheme,
+    onClick,
+    isSelected,
+  } = props;
+  return (
+    <button
+      type="button"
+      aria-current={ariaCurrent ? "page" : false}
```
Comment (gkurck) [inline packages/cyberstorm/src/components/PaginationButton.tsx:29]: Only the current page needs the aria attribute attached - the default is `false` so there's no need to explicitly state that each time.

File: packages/cyberstorm/src/components/PaginationButton.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/PaginationButton.tsx]: The notion about anchor vs button is up for debate, but the notion regarding the ellipsis items still stands.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (gkurck) [reply packages/cyberstorm/src/components/Pagination.tsx]: You nailed the part of wrapping the thing in `<nav>`. Now that `<nav>` is the root of the component, `.root` should be applied there. The the sake of semantics and maintainability `<ul>` and `<li>` should have classes applied as well.

You missed the part about `<ul>` requiring `<li>` as it's children. You can look at [the specs of ul](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul#technical_summary) and check out what's permitted content for it.

**Review by gkurck — CHANGES_REQUESTED** [review]
Some are stylistic/semantic issue to be fixed while some are up for debate, such as the anchor vs button discussion.

File: apps/cyberstorm-storybook/stories/components/PaginationButton.stories.tsx [hunk 32]
```diff
@@ -0,0 +1,17 @@
...
+const meta = {
+  title: "Cyberstorm/Components/PaginationButton",
+  component: PaginationButton,
+} as ComponentMeta<typeof PaginationButton>;
+
+const Template: ComponentStory<typeof PaginationButton> = (args) => (
+  <PaginationButton {...args} />
+);
+
+const DefaultButton = Template.bind({});
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/PaginationButton.stories.tsx:14]: This was more of @VilppeRiskidev vision how we should do this. Argument was that it's better documentation of button itself and shows how the actual it works and what arguments could be given.

[the remark this answers is not in the document]
Comment (VilppeRiskidev) [reply apps/cyberstorm-storybook/stories/components/PaginationButton.stories.tsx:14]: Yup, I think (change my mind) that the storybook's main purpose is documentation. I think having a story for each component documents how to use them. At least it shows the props.
One thing to do (in the "near" future = I'd say right after or even during my current milestone), would be to order the components into their own folders, which I'd also relfect to the storybook.

File: packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css (hunk 30 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css]: We did talk this through and consensus was that this was ok for now iirc?

File: packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+  --bg-color: var(--color-surface--3);
+}
+
+.label {
+  font: var(--font-body);
+  font-weight: var(--font-weight-500);
+  line-height: var(--line-height-l);
+}
+
+.button__default {
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/componentStyles/PaginationButton.module.css:25]: Hard coded this to match primary color for now, there will be later PR to fix this.

File: packages/cyberstorm/src/components/PaginationButton.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/PaginationButton.tsx]: It's now in <span> element.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx]: Did add logic for prev & next buttons to be hidden as described.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx]: Got it done as you mentioned it should be done 👍

File: packages/cyberstorm/src/components/PaginationButton.tsx (hunk 26 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/PaginationButton.tsx:27]: Pointed it out on original comment but this should be corrected now 👍

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 25 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:87]: Good to know! Fixed it that way.

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 24 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:77]: Got your point and changed it to match newPage. Will try my luck on other cases to see where is that line when it's ok 😉

File: packages/cyberstorm/src/components/Pagination.tsx (hunk 22 above)
Comment (Cyanideankst) [reply packages/cyberstorm/src/components/Pagination.tsx:32]: This sense shall be made to be true!

File: apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx [hunk 34]
```diff
@@ -0,0 +1,50 @@
...
+  );
+};
+
+const ReferencePagination = Template.bind({});
+ReferencePagination.args = defaultArgs;
+
+const DefaultPagePagination = Template.bind({});
+DefaultPagePagination.args = {
+  ...defaultArgs,
+  defaultPage: 3,
```
[the remark this answers is not in the document]
Comment (Cyanideankst) [reply apps/cyberstorm-storybook/stories/components/Pagination.stories.tsx:36]: Fixed it!

**Review by VilppeRiskidev — APPROVED** [review]

**Review by VilppeRiskidev — APPROVED** [review]

**Review by gkurck — APPROVED** [review]
Approving so this can be merged.

File: apps/cyberstorm-storybook/stories/components/PaginationButton.stories.tsx (hunk 32 above)
[the remark this answers is not in the document]
Comment (anttimaki) [reply apps/cyberstorm-storybook/stories/components/PaginationButton.stories.tsx:14]: I agree with the idea that documenting the components in Storybook is something we should pursue.

However, on this particular component's case, I disagree.

* The component will probably never be used just by itself, only within the Pagination component (at least the name implies so), so it's documented in Pagination component's story
* There's currently exactly one story. When checking the code, I see the component takes no props nor children. If this is supposed to be documentation, it's IMO quite poor documentation. It tells me nothing about the how the component can be used.
* Visually this renders an empty button that's barely visible on the background. Again, if this is supposed to be documentation, we're currently documenting a use case no one will probably ever want to use

That being said, I acknowledge that this is a low impact issue. However, I think it might work as precedent and affect the shape of our storybook in the future. So even if nothing gets done, I'd like to say for the record that IMO it would be better to either have the story document some viable use cases, or remove the story altogether,

**Review by anttimaki — APPROVED** [review]
I was asked on Discord to *approve* this PR rather than *review* it. So I only marked my earlier comments that *seemed* to be resolved as resolved, and left the rest open so they can be addressed per our regular process.

I **DID NOT REVIEW** the PR after the numerous changes that have been made since my last review.

If I misunderstood, ping me and I'll do a proper review.

**Comment by VilppeRiskidev** [comment]
Merging this, I'll do a round of some cleaning when I use this in my branch.
